### PR TITLE
Makes Meta feature calculation take into account the target class 

### PR DESCRIPTION
**Pull request by janvanrijn**

### CODE REVIEW

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.openml.webapplication.features.CharacterizerFactory;
+import org.openml.webapplication.features.FantailConnector;
+import org.openml.webapplication.testutils.DatasetFactory;
+
+import weka.core.Instances;
+
+public class TestFantailConnector {
+	
+	private static Method makeMethodAccessible() throws NoSuchMethodException, SecurityException {
+		Class[] cArgClasses = new Class[6];
```
Comment (joaquinvanschoren) [inline src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:23]: bad indent

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        cArgClasses[0] = int.class;
+        cArgClasses[1] = Instances.class;
+        cArgClasses[2] = List.class;
+        cArgClasses[3] = List.class;
+        cArgClasses[4] = String.class;
+        cArgClasses[5] = List.class;
+        
+        Method method = FantailConnector.class.getDeclaredMethod("extractFeatures", cArgClasses);
+		method.setAccessible(true);
```
Comment (joaquinvanschoren) [inline src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:32]: bad indent

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		return method;
+	}
+	
+	@Test
+	public final void testXorNumeric() throws Exception {
+		List<Characterizer> characterizers = CharacterizerFactory.all(null);
```
Comment (joaquinvanschoren) [inline src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:38]: bad indent

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        String[] target = {"y"};
+        Object[] cArgValues = new Object[6];
+        cArgValues[0] = -1;
+        cArgValues[1] = DatasetFactory.getXORNumericNoClass();
+        cArgValues[2] = new ArrayList<String>(Arrays.asList(target));
+        cArgValues[3] = null;
+        cArgValues[4] = null;
+        cArgValues[5] = new ArrayList<String>(Arrays.asList(new String[0]));
+		
+		Method method = makeMethodAccessible();
```
Comment (joaquinvanschoren) [inline src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:50]: bad indent

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		
+		Set<Quality> result = (Set<Quality>) method.invoke(connector, cArgValues);
+		
+		assertEquals(CharacterizerFactory.getExpectedQualities(characterizers).size(), result.size());
+	}
+	
+	@Test
+	public final void testXorNominal() throws Exception {
+		List<Characterizer> characterizers = CharacterizerFactory.all(null);
```
Comment (joaquinvanschoren) [inline src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:59]: ...

File: src/main/java/org/openml/webapplication/Main.java [hunk 6]
```diff
@@ -117,21 +118,20 @@ public static void main( String[] args ) {
 					// bootstrap process dataset
 					String processMode = cli.hasOption("x") ? "random" : "normal";
 					new ProcessDataset(apiconnector, id, processMode);
-				} else if( function.equals("extract_features_all") ) {
 					
+				} else if( function.equals("extract_features_all") ) {
```
Comment (joaquinvanschoren) [inline src/main/java/org/openml/webapplication/Main.java:122]: what happened to the option to select a given characterizer? Was this removed??

File: src/main/java/org/openml/webapplication/fantail/dc/statistical/AttributeEntropy.java [hunk 7]
```diff
@@ -69,7 +69,7 @@
 		}
 		
 		Map<String, Double> qualities = new HashMap<String, Double>();
-		if (data.classAttribute().isNominal()) {
+		if (data.classIndex() > 0 && data.classAttribute().isNominal()) {
```
Comment (joaquinvanschoren) [inline src/main/java/org/openml/webapplication/fantail/dc/statistical/AttributeEntropy.java]: Should be >= 0? It is possible that the first feature (id 0) is the target.

File: src/main/java/org/openml/webapplication/features/CharacterizerFactory.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+			String[] repOption = { "-L", "" + i };
+			batchCharacterizers.add(new GenericLandmarker("REPTreeDepth" + i, "weka.classifiers.trees.REPTree", 2, repOption));
+
+			String[] randomtreeOption = { "-depth", "" + i };
+			batchCharacterizers.add(new GenericLandmarker("RandomTreeDepth" + i, "weka.classifiers.trees.RandomTree", 2, randomtreeOption));
+		}
+		return batchCharacterizers;
+	}
+	
+	public static List<Characterizer> simple() {
```
Comment (joaquinvanschoren) [inline src/main/java/org/openml/webapplication/features/CharacterizerFactory.java:47]: Where is this called?

File: src/main/java/org/openml/webapplication/Main.java (hunk 6 above)
Comment (janvanrijn) [reply src/main/java/org/openml/webapplication/Main.java:122]: No, not merged yet.

File: src/main/java/org/openml/webapplication/fantail/dc/statistical/AttributeEntropy.java (hunk 7 above)
Comment (janvanrijn) [reply src/main/java/org/openml/webapplication/fantail/dc/statistical/AttributeEntropy.java]: Nice catch, thanks!

File: src/main/java/org/openml/webapplication/features/CharacterizerFactory.java (hunk 8 above)
Comment (janvanrijn) [reply src/main/java/org/openml/webapplication/features/CharacterizerFactory.java:47]: ProcessDataset.java line 61

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java (hunk 1 above)
Comment (janvanrijn) [reply src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:23]: This one is fine right?

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java (hunk 2 above)
Comment (janvanrijn) [reply src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:32]: Yes, sorry

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java (hunk 3 above)
Comment (janvanrijn) [reply src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:38]: Agreed

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java (hunk 4 above)
Comment (janvanrijn) [reply src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:50]: Agreed

File: src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java (hunk 5 above)
Comment (janvanrijn) [reply src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:59]: Agreed

Comment (janvanrijn) [reply src/test/java/org/openml/webapplication/fantail/TestFantailConnector.java:59]: This one should be fine by now, as the ones below are on the same level

**Review by joaquinvanschoren — APPROVED** [review]

**Pull request merged**